Use spack-stack-1.6.0 environment, in accordance with the current UFS-WM stack

natalie-perlin opened the pull request:
## DESCRIPTION OF CHANGES: 
Modulefiles are updated to use spack-stack-1.6.0 environment 
as currently used by UFS-WM and UFS-SRW App.

The main difference of the software package versions as related to the UFS_UTILS  is an updated  version of g2/3.5.1. It would further enable implement changes to the UFS_UTILS that require this higher version of g2, such as reading large grib2 files.

Files adapted for platforms and compilers, except for Stampede (no longer supported or used by UFS as Tier 1 platforms) and WCOSS2. 

## TESTS CONDUCTED: 

- [x] Compile branch on all Tier 1 machines using lntelllvm/Intel (Orion, Jet, Hera, WCOSS2). Done using 9701aa4.
- [x] Compile branch on Hera using GNU. Done using 9701aa4.
- [x] Compile branch in 'Debug' mode on WCOSS2. Done using 9701aa4.
- [x] Compile with Doxygen on any machine with no errors. Done on WCOSS2 using 9701aa4.
- [x] Run unit tests locally on any Tier 1 machine. Done on Hera using Gnu and 9701aa4.
- [x] Run relevant consistency tests locally on all Tier 1 machines. Done on Orion, Hercules, Jet, Hera and WCOSS2 using 9701aa4. All tests passed.

## DEPENDENCIES:
None.

## DOCUMENTATION:
N/A

## ISSUE: 
This PR is the first step to contribute in resolving the issue https://github.com/ufs-community/ufs-srweather-app/issues/1191 that requires reading large grib2 files.

Commit:
modulefiles to use spack-stack-1.6.0, fms-2024.1 environment that contains g2/3.5.1

Commit:
updated modulefiles/build.hera.gnu.lua

GeorgeGayno-NOAA commented:
@DavidHuber-NOAA - let us know if this is a problem for the global workflow.

GeorgeGayno-NOAA commented:
@natalie-perlin - I only have access to the Tier 1 machines (Hera, Jet, Orion, Hercules and WCOSS2). You will need to test on the other machines.

GeorgeGayno-NOAA commented:
On Hera, when I compile and run the `chgres_cube` consistency tests using Gnu, I get this slurm error:

```
The application appears to have been direct launched using "srun",
but OMPI was not built with SLURM's PMI support and therefore cannot
execute. There are several options for building PMI support under
SLURM, depending upon the SLURM version you are using:

  version 16.05 or later: you can use SLURM's PMIx support. This
  requires that you configure and build SLURM --with-pmix.

  Versions earlier than 16.05: you must use either SLURM's PMI-1 or
  PMI-2 support. SLURM builds PMI-1 by default, or you can manually
  install PMI-2. You must then build Open MPI using --with-pmi pointing
  to the SLURM PMI library location.

Please configure as appropriate and try again.
```

'develop' uses Gnu v9.2. This branch uses v13.3.0.

natalie-perlin commented:
> On Hera, when I compile and run the `chgres_cube` consistency tests using Gnu, I get this slurm error:
> 
> ```
> The application appears to have been direct launched using "srun",
> but OMPI was not built with SLURM's PMI support and therefore cannot
> execute. There are several options for building PMI support under
> SLURM, depending upon the SLURM version you are using:
> 
>   version 16.05 or later: you can use SLURM's PMIx support. This
>   requires that you configure and build SLURM --with-pmix.
> 
>   Versions earlier than 16.05: you must use either SLURM's PMI-1 or
>   PMI-2 support. SLURM builds PMI-1 by default, or you can manually
>   install PMI-2. You must then build Open MPI using --with-pmi pointing
>   to the SLURM PMI library location.
> 
> Please configure as appropriate and try again.
> ```
> 
> 'develop' uses Gnu v9.2. This branch uses v13.3.0.

Will take a look. Some additional env. variables and or/linking flags could be required (as for other repos), looking into this.

Commit:
changes to build.hera.gnu.lua modulefile to fix linking issues and runtime errors

natalie-perlin commented:
@GeorgeGayno-NOAA - updated the build.hera.gnu.lua modulefile, adding cpp flags, library paths and preloaded library to address the runtime issues. 
When building the executables, it is important to open a fresh terminal to avoid gnu module conflicts with any system default gnu modules.

Building the executable was done as following:
```
export target=hera
export compiler=gnu
./build_all.sh >& log.build.hera.gnu13.3. &
```
The log file is attached: l[log.build.hera.gnu13.3.txt](https://github.com/user-attachments/files/19128284/log.build.hera.gnu13.3.txt)

GeorgeGayno-NOAA commented on modulefiles/build.hera.gnu.lua:
```diff
@@ -1,14 +1,16 @@
 
 hpss_ver=os.getenv("hpss_ver") or ""
 load(pathJoin("hpss", hpss_ver))
 
```
I see several references to 'stmp1'. At some point, will these codes be moved to a more permanent location?

natalie-perlin replied:
@GeorgeGayno-NOAA - 
Yes, subsequent versions of spack-stacks use /contrib/spack-stack space. And they will use system-wide installed compiler versions.

natalie-perlin replied:
For example, spack-stack-1.9.0 has the following environment:
`/contrib/spack-stack/spack-stack-1.9.0/envs/ue-gcc-13.2.0`

GeorgeGayno-NOAA replied:
Ok. I hesitate to merge something that points to temporary directories.

natalie-perlin replied:
My apologies for not being more precise in my response...
The directories under /scratch1/ and /scratch2/ (i.e. /scratch2/NCEPDEV/stmp1/role.epic/ , were not considered temporary directories , but the space that was given to EPIC and used during the last few years for building software stacks. There was a recent policy change that EPIC could install software stacks under /contrib/spack-stack/ space, and also the need to move other software to /scratch3 and /scratch4/. So all the subsequent stacks use new locations. 

Until there will be a next decision to move to the spack-stack version (likely 1.9.0 or 1.9.1), the current  gnu-bases stacks in the modulefiles of ufs-wm, ufs-srw, use the software and stack locations as in this current PR. 

https://github.com/ufs-community/ufs-weather-model/blob/develop/modulefiles/ufs_hera.gnu.lua
https://github.com/ufs-community/ufs-weather-model/blob/develop/modulefiles/ufs_hera.gnu.lua

Several other repositories, such as https://github.com/NOAA-EMC/AQM-utils, https://github.com/NOAA-EMC/UPP, or https://github.com/ufs-community/land-DA_workflow do not use gnu as a supported compiler. 

There is a updated location of the gnu/13.3 compiler on /scratch4/NCEPDEV/stmp/role.epic/ and gnu-based spack-stack-1.6.0 environment in  /contrib/spack-stack/spack-stack-1.6.0/envs/gnu-fms-2024.01/. I and not sure whether it been tested with either ufs-weather-model or ufs-srweather-app repository, but has been re-built in the new EPIC locations in the same way as the stacks currently used by the WM and SRW. Gnu13.3 was required to build some of the packages in spack-stack, at the time no needed compilers were available as system modules.

I have updates the modulefile with the new gnu/13.3 and stack location, and have built the executables using the updated build.hera.gnu.lua modulefile. 
Please see the log attached:
[log.build.hera.gnu13.3_new.txt](https://github.com/user-attachments/files/19164571/log.build.hera.gnu13.3_new.txt)

Let me know how you would like to proceed, whether to keep the last commit with the updated locations, or to revert to the stack used by the current ufs-weather-model and ufs-srweather-app.

GeorgeGayno-NOAA replied:
I would be consistent with the current ufs-weather-model and ufs-srweather-app.

natalie-perlin replied:
Thank you. And sure, reversing it to use the same location as current ufs-weather-model and ufs-srw.

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA - updated the build.hera.gnu.lua modulefile, adding cpp flags, library paths and preloaded library to address the runtime issues. When building the executables, it is important to open a fresh terminal to avoid gnu module conflicts with any system default gnu modules.
> 
> The following has been added to the modulefile:
> 
> ```
> prepend_path("CPPFLAGS", " -I/apps/slurm_hera/23.11.3/include/slurm"," ")
> prepend_path("LD_LIBRARY_PATH", "/apps/slurm_hera/23.11.3/lib")
> setenv("LD_PRELOAD", "/scratch2/NCEPDEV/stmp1/role.epic/installs/gnu/13.3.0/lib64/libstdc++.so.6")
> ```
> 
> Building the executable was done as following:
> 
> ```
> export target=hera
> export compiler=gnu
> ./build_all.sh >& log.build.hera.gnu13.3. &
> ```
> 
> The log file is attached: l[log.build.hera.gnu13.3.txt](https://github.com/user-attachments/files/19128284/log.build.hera.gnu13.3.txt)

That fixed it. I was able to run the unit tests and `chgres_cube` consistency tests using Gnu.

DavidHuber-NOAA committed:
Update hsi path on Gaea clusters (#1031)

Fixes #1030.

GeorgeGayno-NOAA committed:
cpld_gridgen - Update regression tests to run in parallel (#1027)

Previously, the tests ran in serial and some parts were run from the command line. This speeds up 
the processing and places all the test functions in the batch queue. The functionality was not changed.

Fixes #1016.

GeorgeGayno-NOAA commented:
@natalie-perlin - be sure to merge in the latest updates from 'develop'.

GeorgeGayno-NOAA commented:
@DavidHuber-NOAA - are these changes OK with the global workflow?

DavidHuber-NOAA reviewed (approved):
@GeorgeGayno-NOAA Yes, these are fine for the global workflow.

Thanks @natalie-perlin for the updates!  Approve.

GeorgeGayno-NOAA commented:
> @DavidHuber-NOAA @GeorgeGayno-NOAA - brought the recent commits in develop into the feature/ss-1.6.0-fms branch, and it is now up-to-date.

Something does not look right with your branch.  I see updates to the `cpld_gridgen` program in this PR. Those updates should be in your branch. When I look at [ss-1.6.0-fms](https://github.com/natalie-perlin/UFS_UTILS/tree/feature/ss-1.6.0-fms) I see a message about your branch being two commits behind the authoritative fork.

natalie-perlin commented:
> > @DavidHuber-NOAA @GeorgeGayno-NOAA - brought the recent commits in develop into the feature/ss-1.6.0-fms branch, and it is now up-to-date.
> 
> Something does not look right with your branch. I see updates to the `cpld_gridgen` program in this PR. Those updates should be in your branch. When I look at [ss-1.6.0-fms](https://github.com/natalie-perlin/UFS_UTILS/tree/feature/ss-1.6.0-fms) I see a message about your branch being two commits behind the authoritative fork.

Looking and fixing it now. Something went wrong when I worked on the other PR.

Commit:
Merge remote-tracking branch 'origin/develop' into feature/ss-1.6.0-fms

GeorgeGayno-NOAA commented:
> > > @DavidHuber-NOAA @GeorgeGayno-NOAA - brought the recent commits in develop into the feature/ss-1.6.0-fms branch, and it is now up-to-date.
> > 
> > 
> > Something does not look right with your branch. I see updates to the `cpld_gridgen` program in this PR. Those updates should be in your branch. When I look at [ss-1.6.0-fms](https://github.com/natalie-perlin/UFS_UTILS/tree/feature/ss-1.6.0-fms) I see a message about your branch being two commits behind the authoritative fork.
> 
> Looking and fixing it now. Something went wrong when I worked on the other PR.

That looks correct now. I will complete some final testing, then merge.
